[flang] Add support for workdistribute construct in flang frontend

skc7 opened the pull request:
This PR introduces wordistribute construct support in flang frontend.
Also adds a workdistribute mlir op.

The work in this PR is C-P and updated from @ivanradanov commit from coexecute implementation:
[flang_workdistribute_iwomp_2024](https://github.com/ivanradanov/llvm-project/commits/flang_workdistribute_iwomp_2024)

skc7 committed:
[flang] Add support for workdistribute construct in flang frontend

skc7 committed:
[OpenMP] Add verifier and tests for workdistribute mlir op.

skc7 committed:
clang-format

mjklemm reviewed (approved):
LGTM

ergawy reviewed (commented):
I think this PR can be split into 2: one for parsing and one for lowering from PFT to MLIR.

Also, we are still missing tests for parsing AFAICT.

Do we have a PR for semantic checks? Or are we planning to do that in a later PR?

skc7 committed:
[mlir][omp] Add more verification for workdistribute op

skc7 committed:
Fix processHostEvalClauses for workdistribute

skc7 committed:
Make workdistribute construct fortran specific.

skc7 committed:
[flang] Add semantics checks for workdistribute construct.

skc7 commented on mlir/test/Dialect/OpenMP/invalid.mlir:
```diff
@@ -3017,3 +3017,26 @@ func.func @invalid_allocate_allocator(%arg0 : memref<i32>) -> () {
+// -----
+func.func @invalid_workdistribute_with_multiple_blocks() -> () {
+  omp.teams {
+  // expected-error @below {{region must contain exactly one block}}
```
Thanks @tblah for feedback. Fixed it. Added more verification in latest patch.

skc7 commented:
> I think this PR can be split into 2: one for parsing and one for lowering from PFT to MLIR.
> 
> Also, we are still missing tests for parsing AFAICT.
> 
> Do we have a PR for semantic checks? Or are we planning to do that in a later PR?

Added semantic checks and tests in the latest patch. Will try to split this PR as suggested. Thanks.

skc7 commented:
This PR was created on my forked repo. Now, moved the branch to llvm repo and also split this PR into 3 parts as @ergawy suggested. 
Part1 : https://github.com/llvm/llvm-project/pull/154376 [OpenMP] Add workdistribute construct in openMP dialect and in llvm frontend
Part2 : https://github.com/llvm/llvm-project/pull/154377 [flang][openmp] Add parser/semantic support for workdistribute
Part3 : https://github.com/llvm/llvm-project/pull/154378 [flang][openmp] Add Lowering to omp mlir for workdistribute construct

skc7 commented:
closing this PR since 3 stacked PRs have been created.
